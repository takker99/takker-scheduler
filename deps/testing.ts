export * from "https://deno.land/std@0.198.0/assert/mod.ts";
