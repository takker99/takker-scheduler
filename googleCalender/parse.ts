import type { Event } from "./events.ts";
import { parse as parseTask } from "../task.ts";
import { addSeconds } from "../deps/date-fns.ts";
import { encodeTitle } from "../lib/utils.ts";

interface ParseInit {
  /** 取得先project */ project: string;
  /** 取得先page title */ title: string;
  /** 行ID */ id: string;
}
/** 指定したテキストから予定と記録のcalendar登録用event objectsを作る
 *
 * @param text - task line (改行不可)
 * @param init - 追加のproperties
 */
export function parse(
  text: string,
  init: ParseInit,
): [Event | undefined, Event | undefined] | undefined {
  const task = parseTask(text);
  if (!task) return;

  const { title, plan, record } = task;
  const source = {
    title: init.title,
    url: `https://scrapbox.io/${init.project}/${
      encodeTitle(init.title)
    }#${init.id}`,
  };

  // eventの名前から[]を外して末尾の日付を消しておく
  const summary = title
    .replace(/\[([^\]]+)\]/g, "$1")
    .trim()
    .replace(/\s\d{4}-\d{2}-\d{2}$/, "")
    // 前に日付があるタイプも消す
    .replace(/^\d{4}-\d{2}-\d{2}\s/, "");

  return [
    plan.start && plan.duration && summary !== ""
      ? {
        summary,
        start: plan.start,
        end: addSeconds(plan.start, plan.duration),
        source,
        id: init.id,
      }
      : // 名無しの予定は飛ばす
        undefined,
    record.start && record.end
      ? {
        summary,
        start: record.start,
        end: record.end,
        source,
        id: init.id,
      }
      : undefined,
  ];
}
