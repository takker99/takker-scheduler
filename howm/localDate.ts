export interface LocalDate {
  year: number;
  month: number;
  date: number;
}
export interface LocalDateTime extends LocalDate {
  hours: number;
  minutes: number;
}

export const isLocalDateTime = (
  date: LocalDate | LocalDateTime,
): date is LocalDateTime => "hours" in date;

export const isBefore = (
  first: LocalDate | LocalDateTime,
  second: LocalDate | LocalDateTime,
): boolean =>
  first.year !== second.year
    ? first.year < second.year
    : first.month !== second.month
    ? first.month < second.month
    : first.date !== second.date
    ? first.date < second.date
    : "hours" in second
    ? "hours" in first
      ? first.hours * 60 + first.minutes < second.hours * 60 + second.minutes
      : 0 < second.hours * 60 + second.minutes
    : false;

export const format = (date: LocalDate | LocalDateTime): string =>
  `${`${date.year}`.padStart(4, "0")}-${zero(date.month)}-${zero(date.date)}${
    "hours" in date ? `T${zero(date.hours)}:${zero(date.minutes)}` : ""
  }`;

const zero = (n: number): string => `${n}`.padStart(2, "0");

export const toDate = (date: LocalDate | LocalDateTime): Date => {
  const str = format(date);
  return new Date(str.includes("T") ? str : `${str}T00:00`);
};

export const fromDate = (date: Date): LocalDateTime => ({
  year: date.getFullYear(),
  month: date.getMonth() + 1,
  date: date.getDate(),
  hours: date.getHours(),
  minutes: date.getMinutes(),
});
