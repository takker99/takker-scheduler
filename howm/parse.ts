import { isValid } from "../deps/date-fns.ts";
import { Result } from "../deps/scrapbox-std.ts";
import {
  format,
  fromDate,
  isBefore,
  LocalDate,
  LocalDateTime,
  toDate,
} from "./localDate.ts";
import { fromStatus, Status, toStatus } from "./status.ts";

export const toFrequency = (
  symbol: string,
): "yearly" | "monthly" | "weekly" | "daily" | undefined => {
  switch (symbol.toLowerCase()) {
    case "y":
      return "yearly";
    case "m":
      return "monthly";
    case "w":
      return "weekly";
    case "d":
      return "daily";
    default:
      return;
  }
};

export interface Repeat {
  type: "yearly" | "monthly" | "weekly" | "daily";
  count: number;
}

/* いずれ実装する
/** 曜日指定情報を加えたDate *
export interface ExDate {
  base: Date;
  startOfDay: 0 | 1 | 2 | 3 | 4 | 5 | 6;
  weekNum: number;
}
*/

export interface Task {
  /** task name */
  name: string;

  /** howm記号 */
  status: Status;

  /** hown記号のオプション */
  speed?: number;

  /** 開始日時 */
  start: LocalDate | LocalDateTime;

  /** 終了日時 */
  end?: LocalDate | LocalDateTime;

  /** 所要時間 (単位はmin) **/
  duration?: number;

  /** 解析前の文字列 */
  raw: string;
}

export interface TaskRangeError {
  name: "TaskRangeError";
  message: string;
}

export interface InvalidDateError {
  name: "InvalidDateError";
  message: string;
}

/** Taskを解析する
 *
 * @param text Taskの文字列
 * @return 解析結果。Taskでなければ`undefined`を返す
 */
export const parse = (
  text: string,
): Result<Task, TaskRangeError | InvalidDateError> | undefined => {
  const matched = text.match(
    /([\+\-!~.])(\d+)?@(\d{4})-(\d{2})-(\d{2})(?:T(\d{2}):(\d{2}))?(?:\/(\d{2}):(\d{2})|\/(?:(?:(?:(\d{4})-)?(\d{2})-)?(\d{2})(?:T(\d{2}):(\d{2}))?)|D(\d+))?/i,
  );
  if (!matched) return;

  const [
    matchedText,
    statusSym,
    speedStr,
    syear,
    smonth,
    sdate,
    shours,
    sminutes,
    ehours2,
    eminutes2,
    eyear,
    emonth,
    edate,
    ehours,
    eminutes,
    durationStr,
  ] = matched;

  /** task name */
  const name = `${text.slice(0, matched.index).trim()}${
    text.slice((matched.index ?? 0) + matchedText.length).trim()
  }`;

  const status_ = toStatus(statusSym);
  // /[\+\-!~.]/以外にマッチするはずがない
  if (!status_) throw Error("status must be +,,-,!,~,.");

  // 後方互換性用コード
  // 先頭が✅か❌のときは、`done`とみなす
  /** task status */
  const status = /^[\uFE00-\uFE0F]*[✅❌]/.test(text) ? "done" : status_;
  const start_: LocalDate = {
    year: parseInt(syear),
    month: parseInt(smonth),
    date: parseInt(sdate),
  };
  /** 開始日時 */
  const start: LocalDate | LocalDateTime = !shours
    ? start_
    : { ...start_, hours: parseInt(shours), minutes: parseInt(sminutes) };
  if (!isValid(toDate(start))) {
    return {
      ok: false,
      value: {
        name: "InvalidDateError",
        message: `The start of the task "${format(start)}" is an invalid date.`,
      },
    };
  }

  const task: Task = { name, status, start, raw: text };
  if (speedStr) task.speed = parseInt(speedStr);
  if (ehours2 || edate) {
    let end: LocalDate | LocalDateTime = {
      year: start.year,
      month: start.month,
      date: start.date,
    };
    if (ehours2) {
      end = { ...end, hours: parseInt(ehours2), minutes: parseInt(eminutes2) };
    } else {
      if (eyear) end.year = parseInt(eyear);
      if (emonth) end.month = parseInt(emonth);
      if (edate) end.date = parseInt(edate);
      if (ehours) {
        end = {
          ...end,
          hours: parseInt(ehours),
          minutes: parseInt(eminutes),
        };
      }
    }
    if (!isValid(toDate(end))) {
      return {
        ok: false,
        value: {
          name: "InvalidDateError",
          message: `The end of the task "${format(end)}" is an invalid date.`,
        },
      };
    }
    if (isBefore(end, start)) {
      return {
        ok: false,
        value: {
          name: "TaskRangeError",
          message: `The start of an task cannot be after its end.\n\nstart:${
            format(start)
          }\nend:${format(end)}`,
        },
      };
    }
    task.end = end;
  }
  if (durationStr) task.duration = parseInt(durationStr);

  return { ok: true, value: task };
};

/** 終日タスクかどうか判定する
 *
 * `start`に時刻が含まれていないタスクは全て終日タスクだとみなす
 */
export const isAllDay = (task: Pick<Task, "start">): boolean =>
  !("hours" in task.start);

/** タスクの所要時間を分単位で得る
 *
 * 所要時間が設定されていない場合は`undefined`を返す
 */
export const getDuration = (
  task: Pick<Task, "start" | "end" | "duration">,
): number | undefined =>
  task.end
    ? isAllDay(task) ? undefined : Math.round(
      (toDate(task.end).getTime() - toDate(task.start).getTime()) / (60 * 1000),
    )
    : task.duration;

/** タスクの終了日時を得る
 *
 * 終日の場合は、最終日の翌日0時を終了日時とする
 */
export const getEnd = (task: Task): LocalDateTime => {
  if (task.end) {
    if ("hours" in task.end) return task.end;
    const end = toDate(task.end);
    end.setDate(end.getDate() + 1);
    return fromDate(end);
  }
  if (task.duration === undefined || !("hours" in task.start)) {
    const end = toDate(task.start);
    end.setHours(0);
    end.setMinutes(0);
    end.setDate(end.getDate() + 1);
    return fromDate(end);
  }
  const end = toDate(task.start);
  end.setMinutes(end.getMinutes() + task.duration);
  return fromDate(end);
};

/** Task objectからタスクリンクの文字列を作る */
export const toString = (task: Task): string => {
  const duration = getDuration(task);

  return `${fromStatus(task.status)}${task.speed ?? ""}@${format(task.start)}${
    duration === undefined ? "" : `D${duration}`
  }${task.name}`;
};
