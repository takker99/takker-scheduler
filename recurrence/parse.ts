import { addMinutes, isValid } from "../deps/date-fns.ts";
import { Result, isNumber } from "../deps/scrapbox-std.ts";
import { format, fromDate, LocalDateTime, toDate } from "../howm/localDate.ts";
import {
    Task,

} from "../howm/parse.ts";
import {
    Event,
} from "../event/parse.ts";
import { Status, toStatus } from "../howm/status.ts";

/** 繰り返しを表す */
export interface Recurrence {
    /** task name */
    name: string;

    /** 開始年 */
    year: number | "*";

    /** 開始月 */
    month: number | "*";

    /** 開始日 */
    date: RegDate

    hours: number;

    minutes: number;

    /** 所要時間 */
    duration: number;

    status?: Status;

    /** 解析前の文字列 */
    raw: string;
}

export type RegDate = number | "*" | "Sun" | "Mon" | "Tue" | "Wed" | "Thu" | "Fri" | "Sat";

export interface LackDateError {
    name: "LackDateError";
    message: string;
}

/** Eventを解析する
 *
 * @param text Eventの文字列
 * @return 解析結果。Eventでなければ`undefined`を返す
 */
export const parse = (
    text: string,
):
    Recurrence | undefined => {
    const matched = text.match(
        /([\+\-!~]?)@(\d{4}|\*)-(\d{2}|\*)-(\d{2}|\*|Sun|Mon|Tue|Wed|Thu|Fri|Sat)T(\d{2}):(\d{2})D(\d+)/i
    );
    // eventでないとき
    if (!matched) return;

    const [
        matchedText,
        statusStr,
        syear,
        smonth,
        sdate,
        shours,
        sminutes,
        durationStr,
    ] = matched;

    /** task name */
    const name = `${text.slice(0, matched.index).trim()}${text.slice((matched.index ?? 0) + matchedText.length).trim()
        }`;
    const status = toStatus(statusStr);
    const year = checkRegNumber(syear);
    const month = checkRegNumber(smonth)
    const date = checkRegDate(sdate)
    const hours = parseInt(shours);
    const minutes = parseInt(sminutes)

    const recurrence: Recurrence = { name, year, month, date, hours, minutes, duration: parseInt(durationStr), raw: text };
    if (status) recurrence.status = status



    return recurrence
};

/** 指定日に繰り返す繰り返しタスクか調べる。
 *
 * 繰り返す場合はそのタスクを指定日の日付で生成する。
 * 繰り返さない場合は`undefined`を返す
 */
export const makeRepeat = (recurrence: Recurrence, date: Date): Task | Event | undefined => {
    const localDate = fromDate(date);
    if (isNumber(recurrence.year) && recurrence.year !== date.getFullYear()) return;
    if (isNumber(recurrence.month) && recurrence.month !== date.getMonth() + 1) return;
    if (isNumber(recurrence.date) && recurrence.date !== date.getDate()) return;
    if (!isNumber(recurrence.date) && recurrence.date !== "*" && toDayNumber(recurrence.date) !== date.getDay()) return;

    return recurrence.status ? {
        name: recurrence.name,
        status: recurrence.status,
        start: localDate,
        duration: recurrence.duration,
        raw: recurrence.raw,
    } : {
        name: recurrence.name,
        start: localDate,
        end: fromDate(addMinutes(date, recurrence.duration)),
        raw: recurrence.raw,
    }
}

/** 数字か`*`のときのみ使える */
const checkRegNumber = (text: string): number | "*" => {
    if (text === "*") return text;
    return parseInt(text);
}

/** エラー処理なし */
const checkRegDate = (text: string): RegDate => {
    switch (text.toLowerCase()) {
        case "*": return "*";
        case "sun": return "Sun"
        case "mon": return "Mon"
        case "tue": return "Tue"
        case "wed": return "Wed"
        case "thu": return "Thu"
        case "fri": return "Fri"
        case "sat": return "Sat"
        default: return parseInt(text)
    }
}

const toDayNumber = (day:
    "Sun" | "Mon" | "Tue" | "Wed" | "Thu" | "Fri" | "Sat"

): 0 | 1 | 2 | 3 | 4 | 5 | 6 => {

    switch (day) {
        case "Sun": return 0
        case "Mon": return 1
        case "Tue": return 2
        case "Wed": return 3
        case "Thu": return 4
        case "Fri": return 5
        case "Sat": return 6
    }
}