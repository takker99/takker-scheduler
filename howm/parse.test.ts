/// <reference lib="deno.ns" />

import { Event, makeRepeat, parse, toString } from "./parse.ts";
import { assertEquals, assertSnapshot } from "../deps/testing.ts";
import { toDate } from "./localDate.ts";

const runTests = async (t: Deno.TestContext, args: string[]) => {
  for (const arg of args) {
    await t.step(arg, () => assertSnapshot(t, parse(arg)));
  }
};

Deno.test("parse()", async (t) => {
  await t.step("compatibility", (t) =>
    runTests(t, [
      "⬜タスクリンク+@2023-09-10",
      "✅終了したリンク+@2023-09-10",
      "❌失敗したタスクも終了扱いとする+@2023-09-10",
      "📝やり途中の+@2023-09-10リンク",
      "先頭にない絵文字は✅タスク判定と関係ない+@2023-09-10",
      "✅日付がないタスクはタスクとみなさない",
    ]));
  {
    const arg = " 前後の 空白は　+@2023-09-11 無視される　  ";
    await t.step(arg, () => assertSnapshot(t, parse(arg)));
  }
  await t.step("type", (t) =>
    runTests(t, [
      "!@2002-10-20 ハイウェイ惑星 〆切",
      "ハイウェイ惑星 〆切!4@2002-10-20",
      "-@2002-10-20 ハイウェイ惑星 買おう",
      "ハイウェイ惑星 買おうかな~@2023-09-12",
      "ハイウェイ惑星+@2002-10-20 買おう",
      "指定子なしはEventとして扱う@2023-04-11T13:00/14:00",
    ]));
  await t.step("start datetime", (t) =>
    runTests(t, [
      "ごはんを食べる+3@2023-04-30",
      "ごはんを食べる+3@2023-04※タスク扱いされない",
      "ごはんを食べる+3@2023※タスク扱いされない",
      "ごはんを食べる+3@2023-04-30T12:23",
      "ごはんを食べる+3@2023-04-30T12:23:45 ※秒は無視する",
      "ごはんを食べる+3@2023-04-30 12:23:※区切り文字が必要",
    ]));
  await t.step("end datetime", (t) =>
    runTests(t, [
      "ごはんを食べる+3@2023-04-13/23D45",
      "ごはんを食べる+3@2023-04-13/05-23D45",
      "ごはんを食べる+3@2023-04-13/2024-05-23D45",
      "ごはんを食べる+3@2023-04-13/2024-05-23T05:47D45",
      "ごはんを食べる+3@2023-04-13/05-23T05:47D45",
      "ごはんを食べる+3@2023-04-13/23T05:47D45",
      "ごはんを食べる+3@2023-04-13/T05:47D45",
      "ごはんを食べる+3@2023-04-13/05:47D45",
      "ごはんを食べる+3@2023-04-13T13:40/23D45",
      "ごはんを食べる+3@2023-04-13T13:40/05-23D45",
      "ごはんを食べる+3@2023-04-13T13:40/2025-05-23D45",
      "ごはんを食べる+3@2023-04-13T13:40/2025-05-23T05:47D45",
      "ごはんを食べる+3@2023-04-13T13:40/05-23T05:47D45",
      "ごはんを食べる+3@2023-04-13T13:40/23T05:47D45",
      "ごはんを食べる+3@2023-04-13T13:40/T05:47D45※終了時刻のみを指定する場合はTをつけない",
      "ごはんを食べる+3@2023-04-13T13:40/14:47D45",
      "ごはんを食べる+3@2023-04-13T13:40/14:47",
      "ごはんを食べる+3@2023-04-13D45/05:47",
    ]));
  await t.step("errors", (t) =>
    runTests(t, [
      "+3@2023-04-30/23D45 終了日時が開始日時より早まってはならない",
      "不正な日時を与えるとInvalid Dateになる+3@2023-04-33/05-23D45",
      "不正な日時+3@2023-04-13T13:40/05:67D45",
    ]));
  await t.step("taskがないときは終了時刻まで明記する", (t) =>
    runTests(t, [
      "ハイウェイ惑星 買おうかな@2023-09-12T12:50D10",
      "ハイウェイ惑星 買おうかな@2023-09-12T12:50/14:20",
      "終了時刻がないのでエラーになる@2023-09-12T12:50",
      "開始時刻がないのでエラー@2023-09-12D10",
    ]));
  await t.step("taskの日付情報から継承する", (t) =>
    runTests(t, [
      "ハイウェイ惑星 買おうかな~@2023-09-12T12:50D10@13",
      "ハイウェイ惑星 買おうかな~@2023-09-12D10@13T12:50",
      "ハイウェイ惑星 買おうかな~@2023-09-12@13T12:50D10",
      "ハイウェイ惑星 買おうかな~@2023-09-12@13T12:50/14:20",
      "ハイウェイ惑星 買おうかな~@2023-09-12@12-13T12:50D10",
      "ハイウェイ惑星 買おうかな~@2023-09-12@2024-12-13T12:50D10",
      "時刻がない~@2023-09-12@13",
      "終了時刻がない~@2023-09-12@13T13:45",
      "開始時刻がない~@2023-09-12@13D45",
      "@13順序を逆にすると認識されない~@2023-09-12T12:50D10",
    ]));
  await t.step("繰り返し", (t) =>
    runTests(t, [
      "ハイウェイ惑星 買おうかな~@2023-09-12T12:50D10R4@13",
      "ハイウェイ惑星 買おうかな~@2023-09-12T12:50D10R47@13",
      "ハイウェイ惑星 買おうかな~@2023-09-12T12:50D10R",
      "ハイウェイ惑星 買おうかな~@2023-09-12T12:50D10RM8",
      "ハイウェイ惑星 買おうかな~@2023-09-12T12:50D10RY8",
      "ハイウェイ惑星 買おうかな~@2023-09-12T12:50D10RW8",
      "ハイウェイ惑星 買おうかな~@2023-09-12T12:50D10RD8",
      "読書会@2023-09-12T12:50D10RD8",
    ]));
});

Deno.test("makeRepeat()", () => {
  {
    const task = "ひるごはん@2023-10-09T12:00D30R1";
    assertEquals(
      toString(
        makeRepeat(
          parse(task)!.value as unknown as Event,
          toDate({ year: 2023, month: 10, date: 10 }),
        )!,
      ),
      "@2023-10-10T12:00D30ひるごはん",
    );
    assertEquals(
      toString(
        makeRepeat(
          parse(task)!.value as unknown as Event,
          toDate({ year: 2023, month: 11, date: 10 }),
        )!,
      ),
      "@2023-11-10T12:00D30ひるごはん",
    );
    assertEquals(
      toString(
        makeRepeat(
          parse(task)!.value as unknown as Event,
          toDate({ year: 2024, month: 11, date: 10 }),
        )!,
      ),
      "@2024-11-10T12:00D30ひるごはん",
    );
  }

  {
    const task = "爪切り@2023-10-09T12:00D5RW1";
    assertEquals(
        makeRepeat(
          parse(task)!.value as unknown as Event,
          toDate({ year: 2023, month: 10, date: 10 }),
        ),
        undefined,
    );
    assertEquals(
        makeRepeat(
          parse(task)!.value as unknown as Event,
          toDate({ year: 2023, month: 10, date: 15 }),
        ),
        undefined,
    );
    assertEquals(
      toString(
        makeRepeat(
          parse(task)!.value as unknown as Event,
          toDate({ year: 2023, month: 10, date: 16 }),
        )!,
      ),
      "@2023-10-16T12:00D5爪切り",
    );
  }
});
