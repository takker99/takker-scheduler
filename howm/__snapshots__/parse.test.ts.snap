export const snapshot = {};

snapshot[`parse() > compatibility 1`] = `
{
  ok: true,
  value: {
    name: "⬜タスクリンク",
    raw: "⬜タスクリンク+@2023-09-10",
    start: {
      date: 10,
      month: 9,
      year: 2023,
    },
    status: "todo",
  },
}
`;

snapshot[`parse() > compatibility 2`] = `
{
  ok: true,
  value: {
    name: "✅終了したリンク",
    raw: "✅終了したリンク+@2023-09-10",
    start: {
      date: 10,
      month: 9,
      year: 2023,
    },
    status: "done",
  },
}
`;

snapshot[`parse() > compatibility 3`] = `
{
  ok: true,
  value: {
    name: "📝やり途中のリンク",
    raw: "📝やり途中の+@2023-09-10リンク",
    start: {
      date: 10,
      month: 9,
      year: 2023,
    },
    status: "todo",
  },
}
`;

snapshot[`parse() > compatibility 4`] = `
{
  ok: true,
  value: {
    name: "先頭にない絵文字は✅タスク判定と関係ない",
    raw: "先頭にない絵文字は✅タスク判定と関係ない+@2023-09-10",
    start: {
      date: 10,
      month: 9,
      year: 2023,
    },
    status: "todo",
  },
}
`;

snapshot[`parse() > compatibility 5`] = `undefined`;

snapshot[`parse() 1`] = `
{
  ok: true,
  value: {
    name: "前後の 空白は無視される",
    raw: " 前後の 空白は　+@2023-09-11 無視される　  ",
    start: {
      date: 11,
      month: 9,
      year: 2023,
    },
    status: "todo",
  },
}
`;

snapshot[`parse() > type 1`] = `
{
  ok: true,
  value: {
    name: "ハイウェイ惑星 〆切",
    raw: "!@2002-10-20 ハイウェイ惑星 〆切",
    start: {
      date: 20,
      month: 10,
      year: 2002,
    },
    status: "deadline",
  },
}
`;

snapshot[`parse() > type 2`] = `
{
  ok: true,
  value: {
    name: "ハイウェイ惑星 〆切",
    raw: "ハイウェイ惑星 〆切!4@2002-10-20",
    speed: 4,
    start: {
      date: 20,
      month: 10,
      year: 2002,
    },
    status: "deadline",
  },
}
`;

snapshot[`parse() > type 3`] = `
{
  ok: true,
  value: {
    name: "ハイウェイ惑星 買おう",
    raw: "-@2002-10-20 ハイウェイ惑星 買おう",
    start: {
      date: 20,
      month: 10,
      year: 2002,
    },
    status: "note",
  },
}
`;

snapshot[`parse() > type 4`] = `
{
  ok: true,
  value: {
    name: "ハイウェイ惑星 買おうかな",
    raw: "ハイウェイ惑星 買おうかな~@2023-09-12",
    start: {
      date: 12,
      month: 9,
      year: 2023,
    },
    status: "up-down",
  },
}
`;

snapshot[`parse() > type 5`] = `
{
  ok: true,
  value: {
    name: "ハイウェイ惑星買おう",
    raw: "ハイウェイ惑星+@2002-10-20 買おう",
    start: {
      date: 20,
      month: 10,
      year: 2002,
    },
    status: "todo",
  },
}
`;

snapshot[`parse() > type 6`] = `undefined`;

snapshot[`parse() > start datetime 1`] = `
{
  ok: true,
  value: {
    name: "ごはんを食べる",
    raw: "ごはんを食べる+3@2023-04-30",
    speed: 3,
    start: {
      date: 30,
      month: 4,
      year: 2023,
    },
    status: "todo",
  },
}
`;

snapshot[`parse() > start datetime 2`] = `undefined`;

snapshot[`parse() > start datetime 3`] = `undefined`;

snapshot[`parse() > start datetime 4`] = `
{
  ok: true,
  value: {
    name: "ごはんを食べる",
    raw: "ごはんを食べる+3@2023-04-30T12:23",
    speed: 3,
    start: {
      date: 30,
      hours: 12,
      minutes: 23,
      month: 4,
      year: 2023,
    },
    status: "todo",
  },
}
`;

snapshot[`parse() > start datetime 5`] = `
{
  ok: true,
  value: {
    name: "ごはんを食べる:45 ※秒は無視する",
    raw: "ごはんを食べる+3@2023-04-30T12:23:45 ※秒は無視する",
    speed: 3,
    start: {
      date: 30,
      hours: 12,
      minutes: 23,
      month: 4,
      year: 2023,
    },
    status: "todo",
  },
}
`;

snapshot[`parse() > start datetime 6`] = `
{
  ok: true,
  value: {
    name: "ごはんを食べる12:23:※区切り文字が必要",
    raw: "ごはんを食べる+3@2023-04-30 12:23:※区切り文字が必要",
    speed: 3,
    start: {
      date: 30,
      month: 4,
      year: 2023,
    },
    status: "todo",
  },
}
`;

snapshot[`parse() > end datetime 1`] = `
{
  ok: true,
  value: {
    end: {
      date: 23,
      month: 4,
      year: 2023,
    },
    name: "ごはんを食べるD45",
    raw: "ごはんを食べる+3@2023-04-13/23D45",
    speed: 3,
    start: {
      date: 13,
      month: 4,
      year: 2023,
    },
    status: "todo",
  },
}
`;

snapshot[`parse() > end datetime 2`] = `
{
  ok: true,
  value: {
    end: {
      date: 23,
      month: 5,
      year: 2023,
    },
    name: "ごはんを食べるD45",
    raw: "ごはんを食べる+3@2023-04-13/05-23D45",
    speed: 3,
    start: {
      date: 13,
      month: 4,
      year: 2023,
    },
    status: "todo",
  },
}
`;

snapshot[`parse() > end datetime 3`] = `
{
  ok: true,
  value: {
    end: {
      date: 23,
      month: 5,
      year: 2024,
    },
    name: "ごはんを食べるD45",
    raw: "ごはんを食べる+3@2023-04-13/2024-05-23D45",
    speed: 3,
    start: {
      date: 13,
      month: 4,
      year: 2023,
    },
    status: "todo",
  },
}
`;

snapshot[`parse() > end datetime 4`] = `
{
  ok: true,
  value: {
    end: {
      date: 23,
      hours: 5,
      minutes: 47,
      month: 5,
      year: 2024,
    },
    name: "ごはんを食べるD45",
    raw: "ごはんを食べる+3@2023-04-13/2024-05-23T05:47D45",
    speed: 3,
    start: {
      date: 13,
      month: 4,
      year: 2023,
    },
    status: "todo",
  },
}
`;

snapshot[`parse() > end datetime 5`] = `
{
  ok: true,
  value: {
    end: {
      date: 23,
      hours: 5,
      minutes: 47,
      month: 5,
      year: 2023,
    },
    name: "ごはんを食べるD45",
    raw: "ごはんを食べる+3@2023-04-13/05-23T05:47D45",
    speed: 3,
    start: {
      date: 13,
      month: 4,
      year: 2023,
    },
    status: "todo",
  },
}
`;

snapshot[`parse() > end datetime 6`] = `
{
  ok: true,
  value: {
    end: {
      date: 23,
      hours: 5,
      minutes: 47,
      month: 4,
      year: 2023,
    },
    name: "ごはんを食べるD45",
    raw: "ごはんを食べる+3@2023-04-13/23T05:47D45",
    speed: 3,
    start: {
      date: 13,
      month: 4,
      year: 2023,
    },
    status: "todo",
  },
}
`;

snapshot[`parse() > end datetime 7`] = `
{
  ok: true,
  value: {
    name: "ごはんを食べる/T05:47D45",
    raw: "ごはんを食べる+3@2023-04-13/T05:47D45",
    speed: 3,
    start: {
      date: 13,
      month: 4,
      year: 2023,
    },
    status: "todo",
  },
}
`;

snapshot[`parse() > end datetime 8`] = `
{
  ok: true,
  value: {
    end: {
      date: 13,
      hours: 5,
      minutes: 47,
      month: 4,
      year: 2023,
    },
    name: "ごはんを食べるD45",
    raw: "ごはんを食べる+3@2023-04-13/05:47D45",
    speed: 3,
    start: {
      date: 13,
      month: 4,
      year: 2023,
    },
    status: "todo",
  },
}
`;

snapshot[`parse() > end datetime 9`] = `
{
  ok: true,
  value: {
    end: {
      date: 23,
      month: 4,
      year: 2023,
    },
    name: "ごはんを食べるD45",
    raw: "ごはんを食べる+3@2023-04-13T13:40/23D45",
    speed: 3,
    start: {
      date: 13,
      hours: 13,
      minutes: 40,
      month: 4,
      year: 2023,
    },
    status: "todo",
  },
}
`;

snapshot[`parse() > end datetime 10`] = `
{
  ok: true,
  value: {
    end: {
      date: 23,
      month: 5,
      year: 2023,
    },
    name: "ごはんを食べるD45",
    raw: "ごはんを食べる+3@2023-04-13T13:40/05-23D45",
    speed: 3,
    start: {
      date: 13,
      hours: 13,
      minutes: 40,
      month: 4,
      year: 2023,
    },
    status: "todo",
  },
}
`;

snapshot[`parse() > end datetime 11`] = `
{
  ok: true,
  value: {
    end: {
      date: 23,
      month: 5,
      year: 2025,
    },
    name: "ごはんを食べるD45",
    raw: "ごはんを食べる+3@2023-04-13T13:40/2025-05-23D45",
    speed: 3,
    start: {
      date: 13,
      hours: 13,
      minutes: 40,
      month: 4,
      year: 2023,
    },
    status: "todo",
  },
}
`;

snapshot[`parse() > end datetime 12`] = `
{
  ok: true,
  value: {
    end: {
      date: 23,
      hours: 5,
      minutes: 47,
      month: 5,
      year: 2025,
    },
    name: "ごはんを食べるD45",
    raw: "ごはんを食べる+3@2023-04-13T13:40/2025-05-23T05:47D45",
    speed: 3,
    start: {
      date: 13,
      hours: 13,
      minutes: 40,
      month: 4,
      year: 2023,
    },
    status: "todo",
  },
}
`;

snapshot[`parse() > end datetime 13`] = `
{
  ok: true,
  value: {
    end: {
      date: 23,
      hours: 5,
      minutes: 47,
      month: 5,
      year: 2023,
    },
    name: "ごはんを食べるD45",
    raw: "ごはんを食べる+3@2023-04-13T13:40/05-23T05:47D45",
    speed: 3,
    start: {
      date: 13,
      hours: 13,
      minutes: 40,
      month: 4,
      year: 2023,
    },
    status: "todo",
  },
}
`;

snapshot[`parse() > end datetime 14`] = `
{
  ok: true,
  value: {
    end: {
      date: 23,
      hours: 5,
      minutes: 47,
      month: 4,
      year: 2023,
    },
    name: "ごはんを食べるD45",
    raw: "ごはんを食べる+3@2023-04-13T13:40/23T05:47D45",
    speed: 3,
    start: {
      date: 13,
      hours: 13,
      minutes: 40,
      month: 4,
      year: 2023,
    },
    status: "todo",
  },
}
`;

snapshot[`parse() > end datetime 15`] = `
{
  ok: true,
  value: {
    name: "ごはんを食べる/T05:47D45",
    raw: "ごはんを食べる+3@2023-04-13T13:40/T05:47D45",
    speed: 3,
    start: {
      date: 13,
      hours: 13,
      minutes: 40,
      month: 4,
      year: 2023,
    },
    status: "todo",
  },
}
`;

snapshot[`parse() > end datetime 16`] = `
{
  ok: true,
  value: {
    end: {
      date: 13,
      hours: 14,
      minutes: 47,
      month: 4,
      year: 2023,
    },
    name: "ごはんを食べるD45",
    raw: "ごはんを食べる+3@2023-04-13T13:40/14:47D45",
    speed: 3,
    start: {
      date: 13,
      hours: 13,
      minutes: 40,
      month: 4,
      year: 2023,
    },
    status: "todo",
  },
}
`;

snapshot[`parse() > end datetime 17`] = `
{
  ok: true,
  value: {
    duration: 45,
    name: "ごはんを食べる/05:47",
    raw: "ごはんを食べる+3@2023-04-13D45/05:47",
    speed: 3,
    start: {
      date: 13,
      month: 4,
      year: 2023,
    },
    status: "todo",
  },
}
`;

snapshot[`parse() > errors 1`] = `
{
  ok: false,
  value: {
    message: "The start of an task cannot be after its end.

start:2023-04-30
end:2023-04-23",
    name: "TaskRangeError",
  },
}
`;

snapshot[`parse() > errors 2`] = `
{
  ok: false,
  value: {
    message: 'The start of the task "2023-04-33" is an invalid date.',
    name: "InvalidDateError",
  },
}
`;

snapshot[`parse() > errors 3`] = `
{
  ok: false,
  value: {
    message: 'The end of the task "2023-04-13T05:67" is an invalid date.',
    name: "InvalidDateError",
  },
}
`;
